### PR TITLE
Allow running some IIS Express variants without publishing

### PR DESCRIPTION
**Pull request by Tratcher**
#1431 Some test variants don't require publishing, especially the defaults Core+Portable and Net+Standalone for x64. x86 net461 and .NET Core Standalone still need to publish.
The average IIS Express test has been reduced from 9s to 3s.

I also changed to providing a config file and site name by default. Most tests do not need to define their own config.

Published tests use their own web.config as before, but unpublished tests embed their settings inside the host.config.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 1]
```diff
@@ -317,20 +388,34 @@ private void ModifyDotNetExePathInWebConfig()
...
-            var config = XDocument.Load(webConfigFile);
+            var config = XDocument.Load(_webConfig);
             var element = config.Descendants("handlers").FirstOrDefault().Descendants("add").FirstOrDefault();
             element.SetAttributeValue(key, value);
-            config.Save(webConfigFile);        
+            config.Save(_webConfig);
+        }
+
+        // These elements are duplicated in the web.config if you publish. Remove them from the host.config.
+        private string RemoveRedundantElements(string serverConfig)
```
Comment (pakrym) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Doesn't IIS merge them?

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: No, it refuses to load the site if you have duplicates.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 2]
```diff
@@ -42,15 +42,79 @@ public IISExpressDeployer(DeploymentParameters deploymentParameters, ILoggerFact
...
-                // For now we always auto-publish. Otherwise we'll have to write our own local web.config for the HttpPlatformHandler
-                DotnetPublish();
+                var targetFramework = DeploymentParameters.TargetFramework;
 
-                var contentRoot = DeploymentParameters.PublishedApplicationRootPath;
+                // IISIntegration uses this layout
+                var dllRoot = Path.Combine(DeploymentParameters.ApplicationPath, "bin", DeploymentParameters.RuntimeArchitecture.ToString(),
+                    DeploymentParameters.Configuration, targetFramework);
+
+                if (!Directory.Exists(dllRoot))
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Add a private function which returns the found dll root and sets PublishApplicationBeforeDeployment.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DeploymentParameters.cs [hunk 3]
```diff
@@ -102,7 +102,7 @@ public DeploymentParameters(TestVariant variant)
 
         public string ServerConfigLocation { get; set; }
 
-        public string SiteName { get; set; }
+        public string SiteName { get; set; } = "HttpTestSite";
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DeploymentParameters.cs]: DefaultTestSite?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Http.config [hunk 4]
```diff
@@ -0,0 +1,1034 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Http.config]: DefaultAppHost.config?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 5]
```diff
@@ -42,15 +42,79 @@ public IISExpressDeployer(DeploymentParameters deploymentParameters, ILoggerFact
...
+                {
+                    // Most repos use this layout
+                    dllRoot = Path.Combine(DeploymentParameters.ApplicationPath, "bin", DeploymentParameters.Configuration, targetFramework);
+
+                    if (!Directory.Exists(dllRoot))
+                    {
+                        // The bits we need weren't pre-compiled, compile on publish
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Where are application published to on pre-compilation? Temp or app local?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 6]
```diff
@@ -42,15 +42,79 @@ public IISExpressDeployer(DeploymentParameters deploymentParameters, ILoggerFact
...
+                        && DeploymentParameters.ApplicationType == ApplicationType.Standalone)
+                {
+                    // Publish is always required to get the correct standalone files in the output directory
+                    DeploymentParameters.PublishApplicationBeforeDeployment = true;
+                }
+
+                var contentRoot = string.Empty;
+                if (DeploymentParameters.PublishApplicationBeforeDeployment)
+                {
+                    DotnetPublish();
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: This will publish to a temp folder. Is that ever cachable between tests that aren't with a fixture?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 7]
```diff
@@ -217,15 +238,74 @@ private async Task<(Uri url, CancellationToken hostExitToken)> StartIISExpressAs
             }
         }
 
-        private string ModifyANCMPathInConfig(string replaceFlag, string dllName, string serverConfig, string contentRoot)
+        private void PrepareConfig(string contentRoot, string dllRoot, int port)
+        {
+            // Config is required. If not present then fall back to one we cary with us.
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: carry

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 8]
```diff
@@ -217,15 +238,74 @@ private async Task<(Uri url, CancellationToken hostExitToken)> StartIISExpressAs
...
+            if (string.IsNullOrWhiteSpace(DeploymentParameters.ServerConfigTemplateContent))
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Let's avoid IsNullOrWhiteSpace, even though it's for testing.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 9]
```diff
@@ -217,15 +238,74 @@ private async Task<(Uri url, CancellationToken hostExitToken)> StartIISExpressAs
...
+            {
+                using (var stream = GetType().Assembly.GetManifestResourceStream("Microsoft.AspNetCore.Server.IntegrationTesting.Http.config"))
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Why not File.ReadAllText?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 10]
```diff
@@ -217,15 +238,74 @@ private async Task<(Uri url, CancellationToken hostExitToken)> StartIISExpressAs
...
+                {
+                    DeploymentParameters.ServerConfigTemplateContent = reader.ReadToEnd();
+                }
+            }
+
+            var serverConfig = DeploymentParameters.ServerConfigTemplateContent;
+
+            // Pass on the applicationhost.config to iis express. With this don't need to pass in the /path /port switches as they are in the applicationHost.config
+            // We take a copy of the original specified applicationHost.Config to prevent modifying the one in the repo.
+            serverConfig = ModifyANCMPathInConfig(replaceFlag: "[ANCMPath]", dllName: "aspnetcore.dll", serverConfig, dllRoot);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Lots of string replaces. I don't know what the performance hit is, but it could be decent. Consider using xml replacement or a StringBuilder?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 11]
```diff
@@ -217,15 +238,74 @@ private async Task<(Uri url, CancellationToken hostExitToken)> StartIISExpressAs
...
+                serverConfig = ReplacePlaceholder(serverConfig, "[HostingModel]", DeploymentParameters.HostingModel.ToString());
+                serverConfig = ReplacePlaceholder(serverConfig, "[AspNetCoreModule]", DeploymentParameters.AncmVersion.ToString());
+            }
+
+            DeploymentParameters.ServerConfigLocation = Path.GetTempFileName();
+            Logger.LogDebug("Saving Config to {configPath}", DeploymentParameters.ServerConfigLocation);
+
+            if (Logger.IsEnabled(LogLevel.Trace))
+            {
+                Logger.LogTrace($"Config File Content:{Environment.NewLine}===START CONFIG==={Environment.NewLine}{{configContent}}{Environment.NewLine}===END CONFIG===", serverConfig);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Can we remove/condense this log message to just show modified components?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 12]
```diff
@@ -42,15 +42,79 @@ public IISExpressDeployer(DeploymentParameters deploymentParameters, ILoggerFact
...
 
-                var contentRoot = DeploymentParameters.PublishedApplicationRootPath;
+                // IISIntegration uses this layout
+                var dllRoot = Path.Combine(DeploymentParameters.ApplicationPath, "bin", DeploymentParameters.RuntimeArchitecture.ToString(),
+                    DeploymentParameters.Configuration, targetFramework);
+
+                if (!Directory.Exists(dllRoot))
+                {
+                    // Most repos use this layout
+                    dllRoot = Path.Combine(DeploymentParameters.ApplicationPath, "bin", DeploymentParameters.Configuration, targetFramework);
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Comment about difference between dllRoot and contentRoot.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DeploymentParameters.cs (hunk 3 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DeploymentParameters.cs]: *shrug*

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs (hunk 5 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Local, and it's not a proper publish, only a compile.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 13]
```diff
@@ -42,15 +42,79 @@ public IISExpressDeployer(DeploymentParameters deploymentParameters, ILoggerFact
...
+                    if (!Directory.Exists(dllRoot))
+                    {
+                        // The bits we need weren't pre-compiled, compile on publish
+                        DeploymentParameters.PublishApplicationBeforeDeployment = true;
+                    }
+                    else if (DeploymentParameters.RuntimeFlavor == RuntimeFlavor.Clr
+                        && DeploymentParameters.RuntimeArchitecture == RuntimeArchitecture.x86)
+                    {
+                        // x64 is the default. Publish to rebuild for the right bitness
+                        DeploymentParameters.PublishApplicationBeforeDeployment = true;
```
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: ANCM is not the restriction here, CLR apps produce an exe with a specific bitness when compiled / published.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs [hunk 14]
```diff
@@ -42,15 +42,79 @@ public IISExpressDeployer(DeploymentParameters deploymentParameters, ILoggerFact
...
+                    }
+                }
+
+                if (DeploymentParameters.RuntimeFlavor == RuntimeFlavor.CoreClr
+                        && DeploymentParameters.ApplicationType == ApplicationType.Standalone)
+                {
+                    // Publish is always required to get the correct standalone files in the output directory
+                    DeploymentParameters.PublishApplicationBeforeDeployment = true;
```
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: They publish to a GUID folder per test, we don't have published content re-use yet. That's pretty high on @pakrym's wishlist.
Also, publish is very scenario specific, e.g. bitness, runtime, etc. so there's limited ability to reuse them anyways.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs (hunk 6 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Not yet.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs (hunk 9 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: It's not a file, it's embedded in the IntegrationTesting assembly. Harder to lose that way.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs (hunk 10 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Yeah, that could use some work. It's not a measurable impact right now compared to the 3 seconds it takes dotnet to start the process 😁. I'm going to punt this re-design to a future PR.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs (hunk 11 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Being tracked as a separate issue.

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Http.config (hunk 4 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Http.config]: *shrug*

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs (hunk 11 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Can you remove the log then 😄

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DotNetCommands.cs [hunk 15]
```diff
@@ -5,7 +5,7 @@
 using System.IO;
 using System.Runtime.InteropServices;
 
-namespace Microsoft.AspNetCore.Server.IntegrationTesting.Common
+namespace Microsoft.AspNetCore.Server.IntegrationTesting
```
Comment (jkotalik) [inline src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DotNetCommands.cs]: Why was this changed?

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs (hunk 11 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/IISExpressDeployer.cs]: Removed

File: src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DotNetCommands.cs (hunk 15 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/DotNetCommands.cs]: I added this class in my last PR and then realized that it didn't match the namespace for anything else in this folder.
